File: src/fr/geobert/radis/service/RadisService.java
package fr.geobert.radis.service;

import android.app.IntentService;
import android.content.Context;
import android.content.Intent;
import android.database.Cursor;
import android.os.PowerManager;
import android.util.Log;
import fr.geobert.radis.MainActivity;
import fr.geobert.radis.RadisConfiguration;
import fr.geobert.radis.data.ScheduledOperation;
import fr.geobert.radis.db.AccountTable;
import fr.geobert.radis.db.OperationTable;
import fr.geobert.radis.db.ScheduledOperationTable;
import fr.geobert.radis.tools.DBPrefsManager;
import fr.geobert.radis.tools.PrefsManager;
import fr.geobert.radis.tools.Tools;

import java.util.HashMap;
import java.util.LinkedHashMap;

public class RadisService extends IntentService {
    private static final String TAG = "RadisService";
    public static final String LOCK_NAME_STATIC = "fr.geobert.radis.StaticLock";
    public static final String CONSOLIDATE_DB = "consolidateDB";
    private static PowerManager.WakeLock lockStatic = null;

    public static void callMe(Context context) {
        RadisService.acquireStaticLock(context);
        context.startService(new Intent(context, RadisService.class));
    }

    public RadisService() {
        super(TAG);
    }

    private void consolidateDbAfterRestore() {
        PrefsManager prefs = PrefsManager.getInstance(this);
        Boolean needConsolidate = prefs.getBoolean(CONSOLIDATE_DB, false);
        Log.d(TAG, "needConsolidate : " + needConsolidate);
        if (needConsolidate) {
            Cursor cursor = AccountTable.fetchAllAccounts(this);
            prefs.put(CONSOLIDATE_DB, false);
            prefs.commit();
            if (cursor != null) {
                if (cursor.moveToFirst()) {
                    do {
                        Log.d(TAG, "CONSOLIDATE ON consolidateDB set to true : " + cursor.getLong(0));
                        AccountTable.consolidateSums(this, cursor.getLong(0));
                    } while (cursor.moveToNext());
                }
                cursor.close();
            }
            MainActivity.refreshAccountList(this);
        }
    }

    @Override
    protected void onHandleIntent(Intent intent) {
        try {
            DBPrefsManager prefs = DBPrefsManager.getInstance(this);
            prefs.fillCache(this);
            processScheduledOps();
        } finally {
            if (getLock(this).isHeld()) {
                Log.d(TAG, "release lock");
                getLock(this).release();
            }
            // If the DB was restored, we consolidate the sums
            consolidateDbAfterRestore();
        }
        stopSelf();
    }

    private void keepGreatestDate(HashMap<Long, Long> greatestDatePerAccount,
                                  final long accountId, final long curOpDate) {
        Long date = greatestDatePerAccount.get(accountId);
        if (null == date) {
            date = (long) 0;
        }
        if (curOpDate > date) {
            greatestDatePerAccount.put(accountId, curOpDate);
        }
    }

//    private class TimeParams {
//        long today = 0;
//        long insertionDate = 0;
//        long currentMonth = 0;
//        long limitInsertionDate = 0;
//    }

//    private TimeParams computeTimeParameters() {
//        //DateFormat formater = Formater.getFullDateFormater(); // used in Log.d
//        GregorianCalendar today = Tools.createClearedCalendar();
//        final long todayInMillis = today.getTimeInMillis();
////        Log.d(TAG, "today : " + Formater.getFullDateFormater().format(today.getTime()));
//
//        GregorianCalendar currentMonth = new GregorianCalendar();
//        currentMonth.setTimeInMillis(todayInMillis);
//        currentMonth.set(Calendar.DAY_OF_MONTH, currentMonth.getActualMaximum(Calendar.DAY_OF_MONTH));
////        Log.d(TAG, "currentMonth : " + formater.format(currentMonth.getTime()));
//
//        GregorianCalendar insertionDate = new GregorianCalendar();
//        final int maxDayOfCurMonth = today.getActualMaximum(Calendar.DAY_OF_MONTH);
//        // manage February if insertionDayOfMonth is 29, 30 or 31
//        final int cfgInsertDay = DBPrefsManager.getInstance(this).getInt(RadisConfiguration.KEY_INSERTION_DATE,
//                Integer.parseInt(RadisConfiguration.DEFAULT_INSERTION_DATE));
//        final int insertionDayOfMonth = cfgInsertDay > maxDayOfCurMonth ? maxDayOfCurMonth : cfgInsertDay;
//        insertionDate.set(Calendar.DAY_OF_MONTH, insertionDayOfMonth);
//        Tools.clearTimeOfCalendar(insertionDate);
////        Log.d(TAG, "insertionDate : " + Formater.getFullDateFormater().format(insertionDate.getTime()));
//
//        if (today.compareTo(insertionDate) > 0) { // what is that for?
//            insertionDate.add(Calendar.MONTH, 1);
////            Log.d(TAG, "adjusted insertionDate : " + Formater.getFullDateFormater().format(insertionDate.getTime()));
//        }
//
//        long insertionDateInMillis = insertionDate.getTimeInMillis();
//        GregorianCalendar limitInsertionDate = new GregorianCalendar();
//        limitInsertionDate.setTimeInMillis(insertionDateInMillis);
//        limitInsertionDate.add(Calendar.MONTH, 1);
////        Log.d(TAG, "limitInsertionDate : " + Formater.getFullDateFormater().format(limitInsertionDate.getTime()));
//
//        final long lastInsertDate =
//                DBPrefsManager.getInstance(this).getLong(RadisConfiguration.KEY_LAST_INSERTION_DATE, 0);
////        Log.d(TAG, "lastInsertDate : " + Formater.getFullDateFormater().format(lastInsertDate));
//        if (lastInsertDate > insertionDateInMillis) {
//            insertionDate.add(Calendar.MONTH, 1);
////            Log.d(TAG, "modified insertionDate : " + Formater.getFullDateFormater().format(insertionDate.getTime()));
//            insertionDateInMillis = insertionDate.getTimeInMillis();
//        }
//        limitInsertionDate.set(Calendar.DAY_OF_MONTH, limitInsertionDate.getActualMaximum(Calendar.DAY_OF_MONTH));
////        Log.d(TAG, "final limitInsertionDate : " + Formater.getFullDateFormater().format(limitInsertionDate.getTime()));
//        TimeParams res = new TimeParams();
//        res.today = todayInMillis;
//        res.currentMonth = currentMonth.getTimeInMillis();
//        res.insertionDate = insertionDateInMillis;
//        res.limitInsertionDate = limitInsertionDate.getTimeInMillis();
//        return res;
//    }

    private synchronized void processScheduledOps() {
        Cursor schOpsCursor = ScheduledOperationTable.fetchAllScheduledOps(this);
        if (schOpsCursor.isFirst()) {
//            DateFormat formater = Formater.getFullDateFormater(); // used in Log.d
            TimeParams timeParams = TimeParams.object$.computeTimeParams(this);

            HashMap<Long, Long> sumsPerAccount = new LinkedHashMap<Long, Long>();
            HashMap<Long, Long> greatestDatePerAccount = new LinkedHashMap<Long, Long>();
            ScheduledOperation op;
            do {
                final long OP_ROW_ID = schOpsCursor.getLong(schOpsCursor.getColumnIndex("_id"));
                op = new ScheduledOperation(schOpsCursor);
                final Long accountId = op.mAccountId;
                final Long transId = op.mTransferAccountId;
                long sum = 0;
                boolean needUpdate = false;

                Cursor accountCursor = AccountTable.fetchAccount(this, accountId);
//                if (null != accountCursor) { // IntelliJ tells me it is always true
                if (!accountCursor.moveToFirst()) {
                    schOpsCursor.moveToNext();
                    continue;
                } else {
                    AccountTable.initProjectionDate(accountCursor);
                }
                accountCursor.close();
//                } else {
//                    schOpsCursor.moveToNext();
//                    continue;
//                }

                // insert all scheduled of the past until current month
//                Log.d(TAG, "insert all scheduled of the past until current month");
//                int i = 0; // for logging purpose
                while (op.getDate() <= timeParams.getCurrentMonth() && !op.isObsolete()) {
                    long opSum = insertSchOp(op, OP_ROW_ID);
//                    i++;
                    if (opSum != 0) {
                        sum = sum + opSum;
                        needUpdate = true;
                    }
                }
//                Log.d(TAG, "inserted " + i + " past scheduled op until current month");
//                i = 0;
                if (timeParams.getToday() >= timeParams.getInsertionDate()) { // it's time to insert scheduled ops
                    while (op.getDate() < timeParams.getLimitInsertionDate() && !op.isObsolete()) {
                        keepGreatestDate(greatestDatePerAccount, accountId, op.getDate());
                        if (transId > 0) {
                            keepGreatestDate(greatestDatePerAccount, transId, op.getDate());
                        }
//                        Log.d(TAG, "op month before : " + op.getMonth());
                        long opSum = insertSchOp(op, OP_ROW_ID);
//                        Log.d(TAG, "op month after : " + op.getMonth());
//                        i++;
                        if (opSum != 0) {
                            sum = sum + opSum;
                            needUpdate = true;
                        }
                    }
//                    Log.d(TAG, "inserted " + i + " ops current month scheduled op");
                }
                ScheduledOperationTable.updateScheduledOp(this, OP_ROW_ID, op, false);
                if (needUpdate) {
                    Long curSum = sumsPerAccount.get(accountId);
                    if (curSum == null) {
                        curSum = (long) 0;
                    }
                    sumsPerAccount.put(accountId, curSum + sum);
                    keepGreatestDate(greatestDatePerAccount, accountId, op.getDate());
                    // the sch op is a transfert, update the dst account sum with -sum
                    if (transId > 0) {
                        curSum = sumsPerAccount.get(transId);
                        if (curSum == null) {
                            curSum = (long) 0;
                        }
                        sumsPerAccount.put(transId, curSum - sum);
                        keepGreatestDate(greatestDatePerAccount, transId, op.getDate());
                    }
                }
            } while (schOpsCursor.moveToNext());

            boolean needUpdate = false;
            Long[] accountIds = sumsPerAccount.keySet().toArray(new Long[sumsPerAccount.size()]);
            for (HashMap.Entry<Long, Long> e : sumsPerAccount.entrySet()) {
                needUpdate = true;
                updateAccountSum(e.getValue(), 0, e.getKey(),
                        greatestDatePerAccount.get(e.getKey()), this);
            }
//            Log.d(TAG, "DOES NEED UPDATE : " + needUpdate);
            if (needUpdate) {
                Intent i = new Intent(Tools.INTENT_REFRESH_NEEDED);
                i.putExtra("accountIds", accountIds);
//                Log.d(TAG, "sendOrderedBroadcast" + i.toString());
                sendOrderedBroadcast(i, null);
            }
//            Log.d(TAG, "save LAST_INSERT_DATE is todayInMillis: " + Formater.getFullDateFormater().format(p.today));
            DBPrefsManager.getInstance(this).put(RadisConfiguration.KEY_LAST_INSERTION_DATE, timeParams.getToday());
        }
        schOpsCursor.close();
    }

    public static void updateAccountSum(final long opSum, final long oldSum, final long accountId,
                                        final long opDate, final Context ctx) {
        AccountTable.updateProjection(ctx, accountId, opSum, oldSum, opDate, -2);
    }

    private long insertSchOp(ScheduledOperation op, final long opRowId) {
        final long accountId = op.mAccountId;
        op.mScheduledId = opRowId;
        boolean needUpdate = OperationTable.createOp(this, op, accountId, false) > -1;
        ScheduledOperation.addPeriodicityToDate(op);
        return needUpdate ? op.mSum : 0;
    }

    public static void acquireStaticLock(Context context) {
        Log.d(TAG, "acquireStaticLock");
        getLock(context).acquire();
    }

    synchronized private static PowerManager.WakeLock getLock(Context context) {
        if (lockStatic == null) {
            PowerManager mgr = (PowerManager) context.getSystemService(Context.POWER_SERVICE);
            lockStatic = mgr.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, LOCK_NAME_STATIC);
            lockStatic.setReferenceCounted(true);
        }
        return lockStatic;
    }

}

File: app/src/main/kotlin/fr/geobert/radis/service/RadisService.kt
package fr.geobert.radis.service

import android.content.Context
import android.content.Intent
import android.os.PowerManager
import android.util.Log
import fr.geobert.radis.MainActivity
import fr.geobert.radis.data.Account
import fr.geobert.radis.data.AccountConfig
import fr.geobert.radis.data.ScheduledOperation
import fr.geobert.radis.db.AccountTable
import fr.geobert.radis.db.OperationTable
import fr.geobert.radis.db.PreferenceTable
import fr.geobert.radis.db.ScheduledOperationTable
import fr.geobert.radis.tools.DBPrefsManager
import fr.geobert.radis.tools.Tools
import fr.geobert.radis.tools.formatDate
import fr.geobert.radis.ui.ConfigFragment
import java.util.*

public class RadisService : android.app.IntentService(RadisService.TAG) {

    private fun consolidateDbAfterRestore() {
        val prefs = DBPrefsManager.getInstance(this)
        val needConsolidate = prefs.getBoolean(CONSOLIDATE_DB, false)
        if (needConsolidate) {
            val cursor = AccountTable.fetchAllAccounts(this)
            prefs.resetCache()
            prefs.fillCache(this)
            if (cursor.moveToFirst()) {
                do {
                    AccountTable.consolidateSums(this, cursor.getLong(0))
                } while (cursor.moveToNext())
            }
            cursor.close()

            MainActivity.Companion.refreshAccountList(this)
        }
    }

    override fun onHandleIntent(intent: android.content.Intent) {
        try {
            val prefs = DBPrefsManager.getInstance(this)
            prefs.fillCache(this)
            processScheduledOps()
        } finally {
            if (getLock(this).isHeld) {
                android.util.Log.d(TAG, "release lock")
                getLock(this).release()
            }
            // If the DB was restored, we consolidate the sums
            consolidateDbAfterRestore()
        }
        stopSelf()
    }

    private fun keepGreatestDate(greatestDatePerAccount: java.util.HashMap<Long, Long>, accountId: Long, curOpDate: Long) {
        val date: Long = greatestDatePerAccount.get(accountId) ?: 0L
        if (curOpDate > date) {
            greatestDatePerAccount.put(accountId, curOpDate)
        }
    }


    private fun processScheduledOps() {
        synchronized(this, {
            val schOpsCursor = ScheduledOperationTable.fetchAllScheduledOps(this)
            if (schOpsCursor.isFirst) {
                val sumsPerAccount = LinkedHashMap<Long, Long>()
                val greatestDatePerAccount = LinkedHashMap<Long, Long>()

                // TODO convert to more a Kotlin way with a map on the cursor
                do {
                    val OP_ROW_ID = schOpsCursor.getLong(schOpsCursor.getColumnIndex("_id"))
                    val op = ScheduledOperation(schOpsCursor)
                    val accountId = op.mAccountId
                    val transId = op.mTransferAccountId
                    var sum: Long = 0
                    var needUpdate = false

                    val accountCursor = AccountTable.fetchAccount(this, accountId)
                    val configCursor = PreferenceTable.fetchPrefForAccount(this, accountId)
                    if (!accountCursor.moveToFirst()) {
                        schOpsCursor.moveToNext()
                        continue
                    } else {
                        AccountTable.initProjectionDate(accountCursor)
                    }
                    val account = Account(accountCursor)
                    val config = AccountConfig(configCursor)

                    accountCursor.close()
                    configCursor.close()

                    val timeParams = TimeParams.computeTimeParams(this, account, config)

                    Log.d("RadisService", "after compute timeparams timeParams.today: ${timeParams.today.formatDate()} / timeParams.insertionDate: ${timeParams.insertionDate.formatDate()}")

                    // insert all scheduled of the past until current month
                    while (op.getDate() <= timeParams.currentMonth && !op.isObsolete()) {
                        Log.d("RadisService", "insert past sch op until current month, op: ${op.getDate().formatDate()} / limit: ${timeParams.currentMonth.formatDate()}")
                        val opSum = insertSchOp(op, OP_ROW_ID)
                        if (opSum != 0L) {
                            sum += opSum
                            needUpdate = true
                        }
                    }

                    Log.d("RadisService", "processScheduledOps timeParams.today: ${timeParams.today.formatDate()} / timeParams.insertionDate: ${timeParams.insertionDate.formatDate()}")
                    if (timeParams.today >= timeParams.insertionDate) {
                        // it's time to insert scheduled ops
                        while (op.getDate() < timeParams.limitInsertionDate && !op.isObsolete()) {
                            keepGreatestDate(greatestDatePerAccount, accountId, op.getDate())
                            if (transId > 0) {
                                keepGreatestDate(greatestDatePerAccount, transId, op.getDate())
                            }
                            val opSum = insertSchOp(op, OP_ROW_ID)
                            if (opSum != 0L) {
                                sum += opSum
                                needUpdate = true
                            }
                        }
                    }
                    ScheduledOperationTable.updateScheduledOp(this, OP_ROW_ID, op, false)
                    if (needUpdate) {
                        val curSum: Long = sumsPerAccount.get(accountId) ?: 0L
                        sumsPerAccount.put(accountId, curSum + sum)
                        keepGreatestDate(greatestDatePerAccount, accountId, op.getDate())
                        // the sch op is a transfert, update the dst account sum with -sum
                        if (transId > 0) {
                            val curSum2 = sumsPerAccount.get(transId) ?: 0L
                            sumsPerAccount.put(transId, curSum2 - sum)
                            keepGreatestDate(greatestDatePerAccount, transId, op.getDate())
                        }
                    }
                    if (config.overrideInsertDate) {
                        updateAccountLastInsertDate(account.id, timeParams.today)
                    }
                } while (schOpsCursor.moveToNext())

                sumsPerAccount.entries.forEach {
                    updateAccountSum(it.value, 0, it.key, greatestDatePerAccount.get(it.key) as Long)
                }

                if (sumsPerAccount.entries.count() > 0) {
                    sendOrderedBroadcast(Intent(Tools.INTENT_REFRESH_NEEDED), null)
                }

                DBPrefsManager.getInstance(this).put(ConfigFragment.KEY_LAST_INSERTION_DATE,
                        Tools.createClearedCalendar().timeInMillis)
            }
            schOpsCursor.close()
        })
    }


    private fun insertSchOp(op: fr.geobert.radis.data.ScheduledOperation, opRowId: Long): Long {
        val accountId = op.mAccountId
        op.mScheduledId = opRowId
        val needUpdate = OperationTable.createOp(this, op, accountId, false) > -1
        ScheduledOperation.Companion.addPeriodicityToDate(op)
        return if (needUpdate) op.mSum else 0
    }

    private fun updateAccountLastInsertDate(accountId: Long, date: Long) {
        AccountTable.updateAccountLastInsertDate(this, accountId, date)
    }

    public fun updateAccountSum(opSum: Long, oldSum: Long, accountId: Long, opDate: Long) {
        AccountTable.updateProjection(this, accountId, opSum, oldSum, opDate, (-2).toLong())
    }

    companion object {
        private val TAG = "RadisService"
        public val LOCK_NAME_STATIC: String = "fr.geobert.radis.StaticLock"
        public val CONSOLIDATE_DB: String = "consolidateDB"
        private var lockStatic: android.os.PowerManager.WakeLock? = null

        public fun callMe(context: android.content.Context) {
            RadisService.acquireStaticLock(context)
            context.startService(android.content.Intent(context, RadisService::class.java))
        }

        public fun acquireStaticLock(context: android.content.Context): PowerManager.WakeLock {
            android.util.Log.d(TAG, "acquireStaticLock")
            val l = getLock(context)
            l.acquire()
            return l
        }

        private fun getLock(context: Context): PowerManager.WakeLock {
            synchronized(this, {
                val lock = if (lockStatic == null) {
                    val mgr = context.getSystemService(Context.POWER_SERVICE) as PowerManager
                    val l = mgr.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, LOCK_NAME_STATIC)
                    l.setReferenceCounted(true)
                    lockStatic = l
                    l
                } else {
                    lockStatic
                }
                return lock as PowerManager.WakeLock
            })
        }
    }

}
